iree/turbine/kernel/wave/workgroup_reordering.py

```diff
@@ -0,0 +1,28 @@
+from .._support.tracing import CapturedTrace
+from .._support.indexing import *
+from ..ops.wave_ops import *
+from ..lang.global_symbols import *
+from .constraints import *
+from .utils.symbol_utils import *
+
+
+def reorder_workgroups(graph: CapturedTrace, reordering_constraints):
+    if len(reordering_constraints) == 0:
+        return
+    wg_subs = {}
+    for c in reordering_constraints:
+        wg_subs[c.wg_dim] = c.reordered_equation
+
+    graph_nodes = graph.walk()
+    ops_with_implicit_index = {"iterate", "get_result"}
+    for node in graph_nodes:
+        custom_node = get_custom(node)
+        if custom_node.index and custom_node.name not in ops_with_implicit_index:
+            for dim, symb_exp in node.index.items():
+                symbols = symb_exp.start.free_symbols
+                for symbol in symbols:
```

Comment:
You can't modify `symb_exp` object inplace as it can be shared between multiple nodes you need to make a copy, also you can avoid most of these checks by just doing `symb_exp = symb_exp.subs(wg_subs, simultaneous=True)`, `simultaneous` change is here https://github.com/iree-org/iree-turbine/pull/859/files#diff-0f91a6de33b5f0b6427f923fa8b78f07bdc94a04bd5b9c7e61cd8099b2a9e346R451-R454